lib/llm/src/kv_router/approx.rs

```diff
@@ -0,0 +1,542 @@
+    }
+
+    /// Returns the expiration Instant for a given key, if it exists.
+    pub fn get_expiry(&self, key: &K) -> Option<&Instant> {
+        self.timers.get(key)
+    }
+
+    /// Polls for expired timers and returns a list of keys for all timers
+    /// that have expired up to the current moment.
+    pub fn poll_expired(&mut self) -> Vec<K> {
+        let mut expired_keys = Vec::new();
+        let now = Instant::now();
+
+        while let Some(Reverse((expiry_time, _))) = self.expirations.peek() {
+            // If the next timer in the heap is not yet expired, we can stop.
+            if *expiry_time > now {
+                break;
+            }
+
+            // The timer might be expired, so pop it from the heap.
+            // We can safely unwrap because we just peeked.
+            let Reverse((expiry_time, key)) = self.expirations.pop().unwrap();
+
+            // CRUCIAL STEP: Check if the popped timer is stale.
+            // A timer is stale if its key is no longer in our authoritative map,
+            // or if the expiration time in the map is different (i.e., it was updated).
+            match self.timers.get(&key) {
+                Some(authoritative_expiry) if *authoritative_expiry == expiry_time => {
+                    // This is a valid, non-stale, expired timer.
+                    // Remove it from the map and add its key to our results.
+                    self.timers.remove(&key);
+                    expired_keys.push(key);
+                }
+                _ => {
+                    // This entry in the heap was stale. It was either removed
+                    // or updated with a new time. We just ignore it and continue.
+                }
+            }
+        }
+
+        expired_keys
+    }
+}
+
+pub struct ApproxKvIndexer {
+    /// A `CancellationToken` for managing shutdown.
+    cancel: CancellationToken,
+    /// A sender for `MatchRequest`s.
+    match_tx: mpsc::Sender<MatchRequest>,
+    /// A sender for `RouterResult`s.
+    route_tx: mpsc::Sender<RouterResult>,
+    /// A sender for remove worker requests.
+    remove_worker_tx: mpsc::Sender<WorkerId>,
+    /// A handle to the background task managing the KV store.
+    task: OnceLock<std::thread::JoinHandle<()>>,
+    /// The size of the KV block this indexer can handle.
+    kv_block_size: usize,
+}
+
+impl ApproxKvIndexer {
+    pub fn new(token: CancellationToken, kv_block_size: usize, ttl: Duration) -> Self {
+        let (match_tx, mut match_rx) = mpsc::channel::<MatchRequest>(2048);
+        let (route_tx, mut route_rx) = mpsc::channel::<RouterResult>(2048);
+        let (remove_worker_tx, mut remove_worker_rx) = mpsc::channel::<WorkerId>(16);
+        let cancel_clone = token.clone();
+        let task = std::thread::spawn(move || {
+            // create a new tokio runtime which will only perform work on a single thread
+            let runtime = tokio::runtime::Builder::new_multi_thread()
```

Comment:
there is a use_current_thread which might be better suited here since you are already on a new thread.